Fix types 🌋

lluia opened the pull request:
<img src="https://media.giphy.com/media/3q1yvRyvZaDF8waQ8I/giphy.gif" width="400" />

This brings **two important changes** to the library regarding types:

1. In `package.son` → `devDependencies` ...

<img src="https://user-images.githubusercontent.com/5938217/68474178-0dfdbe00-0225-11ea-8179-8b5a7ad4caa9.png" width="350" />

2. In `tsconfig.json` ...

<img src="https://user-images.githubusercontent.com/5938217/68474292-49988800-0225-11ea-8226-c271f0b727eb.png" width="240" />

We didn't had the types for `styled-components` because they had a long standing issue of adding [`@types/react-native` when installed](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/33311), which produced all sorts of weird type errors... 🙄

This is being sorted now and we're now able to bring back `--noImplicitAny`, which in turn raised a load of errors I had to fix. 

 It's now ready to be reviewed, I've tried it in **Zopa Savings** and I'm confident the changes are stable enough 🤞🏻

poteirard commented on src/components/atoms/Button/Button.tsx:
```diff
@@ -99,33 +101,34 @@ const SButton = styled.button<IButtonProps>`
-    color: ${({ styling }) => hoverFontColors[styling]};
-    background-color: ${({ styling }) => activeBackgroundColors[styling]};
+    color: ${({ styling = 'primary' }) => hoverFontColors[styling]};
+    background-color: ${({ styling = 'primary' }) => activeBackgroundColors[styling]};
```
I think it will look cleaner to have
```ts
SButton.defaultProps = {
  styling: 'primary',
  sizing: 'default',
}
```
So we don't have to repeat that login in the props func all the time

lluia replied:
Good point, I think `defaultProps` were previously declared on the wrapper and not on the _styled component_, hence TS wasn't able to infer them. I'll update it 👌

poteirard commented on src/components/atoms/Link/Link.tsx:
```diff
@@ -1,15 +1,20 @@
-  color?: IBaseColorSpec['secondary'] | INeutralColorSpec['white'];
+export interface ILinkProps
+  extends React.AnchorHTMLAttributes<HTMLAnchorElement>,
+    React.RefAttributes<HTMLAnchorElement> {
```
You can use this instead of those 2 extensions: `React.DetailedHTMLProps<React.AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>`
Actually I think we should be using always this. As this is the return type of `styled.a`

lluia replied:
I think we still need to reference `React.RefAttributes` in the type somehow, otherwise TS complains about `ref` below 😪

poteirard commented on src/components/atoms/Link/Link.tsx:
```diff
@@ -42,12 +47,13 @@ const TargetIcon = (props: ITargetIconProps) => {
-const Link: FC<ILinkProps> = React.forwardRef<HTMLAnchorElement, ILinkProps>((props, ref) => {
-  const { children, color = colors.base.secondary, ...rest } = props;
+const Link = React.forwardRef<HTMLAnchorElement, ILinkProps>((props, ref) => {
+  const { children, color = colors.base.secondary, target, ...rest } = props;
```
this line is not needed:
```
const Link = React.forwardRef<HTMLAnchorElement, ILinkProps>(({ children, color = colors.base.secondary, target, ...rest }, ref) => {
```

poteirard commented on src/components/hooks/useForm/useForm.ts:
```diff
@@ -16,8 +14,8 @@ export interface IFieldProps {
-  onChange: (e: any) => void;
-  onBlur: () => void;
+  onChange: (eventData: unknown) => void;
+  onBlur: (eventData: unknown) => void;
```
maybe a better type would be:
```
  onChange: React.ChangeEventHandler<Element>;
  onBlur: React.FocusEventHandler<Element>;
```

lluia replied:
It would, however in this implementation `onChange` and `onBlur` not always expect the actual event object, sometimes [they receive the value directly](https://github.com/zopaUK/react-components/blob/master/src/components/organisms/Form/FormRadioField/FormRadioField.tsx#L14) ( cc. @thegrinder  ) hence I typed it as `unknown`.

I'll try to refactor the `<Form.* />` components so that they all send the actual event to this handlers 💆🏻‍♂️

poteirard commented on src/components/icons/Chevron/Chevron.tsx:
```diff
@@ -16,14 +16,17 @@ export interface IChevronProps extends React.SVGProps<SVGSVGElement> {
+
+  if (typeof direction === 'number') return direction.toString();
+
+  return directionMap[direction];
```
I think it looks better like this:
```ts
return typeof direction === 'number' ? direction.toString() : directionMap[direction];`
```
Also I think we should avoid having conditionals in on line but that's another topic

lluia replied:
ha ha, I ❤️ one line conditionals 💅🏻 😛 , can't see nothing wrong with them, seems a bit biased as some people think they don't read good and other like me think they read better 💆🏻‍♂️

poteirard commented on src/components/layout/FlexCol/FlexCol.tsx:
```diff
@@ -42,15 +42,19 @@ const genAutoWidth = breakpoint => css`
 
-const getWidth = (breakpoint, colWidth, cols) => {
+const getWidth = (breakpoint: TGridBreakpoints, colWidth: TColWidth, cols: IFlexColProps['cols']) => {
   if (colWidth === 'hidden') {
```
I know it's not on this scope but I think this would be better like this:

```ts
const getWidth = (breakpoint: TGridBreakpoints, colWidth: TColWidth, cols: IFlexColProps['cols']) => {
  switch (colWidth) {
    case 'hidden':
      return genHidden(breakpoint);
    case 'auto':
      return genAutoWidth(breakpoint);
    case 'fill':
      return genFillWidth(breakpoint);
    default:
      return genRelativeWidth(breakpoint, colWidth, cols);
  }
};

```

lluia replied:
Yup don't mind to try to refactor it 👍🏻

poteirard commented on src/components/organisms/Navbar/NavbarLink/NavbarLink.tsx:
```diff
@@ -2,11 +2,11 @@ import React, { FC } from 'react';
 export interface IStyledNavbarLinkProps extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
   active: boolean;
-  color: string;
+  color?: ILinkProps['color'];
```
Why not `IStyledNavbarLinkProps extends ILinkProps` ? and then you can get rid of this

lluia replied:
Hmm good point 🤔 , I'll give it a go 👍🏻

poteirard commented on src/constants/grid.ts:
```diff
@@ -1,4 +1,14 @@
+export type TGridWidths = 's' | 'm' | 'l' | 'xl';
+
+interface IGrid {
+  breakpoints: { [k in TGridBreakpoints]: number };
```
`breakpoints:  Record<TGridBreakpoints, number>;`

poteirard commented on src/constants/grid.ts:
```diff
@@ -1,4 +1,14 @@
+  breakpoints: { [k in TGridBreakpoints]: number };
+  cols: number;
+  gutter: number;
+  width: { [k in TGridWidths]: number };
```
`breakpoints:  Record<TGridWidths, number>;`

poteirard commented on src/helpers/responsiveness.ts:
```diff
@@ -11,22 +9,24 @@ const sizes = {
+const maxMedia = {
+  phone: interpolate('phone', 'max'),
+  tablet: interpolate('tablet', 'max'),
+  desktop: interpolate('desktop', 'max'),
```
Good refactor :)

lluia commented:
@poteirard addressed your suggestions 💅🏻 👍🏻  apart from the one referring to `<Link />` type extension ( _see my comment_ ), could you have a second look 🧐 ? 🙏🏻

poteirard commented:
:tada: This PR is included in version 2.5.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@zopauk/react-components)
- [GitHub release](https://github.com/zopaUK/react-components/releases/tag/v2.5.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
